ospf6d/ospf6_asbr.c

```diff
@@ -2613,3 +2693,980 @@ void install_element_ospf6_debug_asbr(void)
+
+	if (IS_OSPF6_DEBUG_AGGR)
+		zlog_debug("%s: Aggr LSA ID: %d flags %x.",
+		   __func__, aggr->id, aggr->aggrflags);
+	/* Dont originate external LSA,
+	 * If it is configured not to advertise.
+	 */
+	if (CHECK_FLAG(aggr->aggrflags, OSPF6_EXTERNAL_AGGRT_NO_ADVERTISE)) {
+		/* If it is already originated as external LSA,
+		 * But, it is configured not to advertise then
+		 * flush the originated external lsa.
+		 */
+		if (lsa) {
+			if (IS_OSPF6_DEBUG_AGGR)
+				zlog_debug("%s: Purge the external LSA %s.",
+					   __func__, lsa->name);
+			ospf6_external_lsa_purge(ospf6, lsa);
+			info->id = 0;
+			rt->path.origin.id = 0;
+		}
+
+		if (aggr_lsa) {
+			if (IS_OSPF6_DEBUG_AGGR)
+				zlog_debug("%s: Purge the aggr external LSA %s.",
+					   __func__, lsa->name);
+			ospf6_asbr_summary_remove_lsa_and_route(ospf6, aggr);
+		}
+
+		UNSET_FLAG(aggr->aggrflags, OSPF6_EXTERNAL_AGGRT_ORIGINATED);
+
+		if (IS_OSPF6_DEBUG_AGGR)
+			zlog_debug("%s: Don't originate the summary address,It is configured to not-advertise.",
+				__func__);
+		return;
+	}
+
+	/* Summary route already originated,
+	 * So, Do nothing.
+	 */
+	if (CHECK_FLAG(aggr->aggrflags, OSPF6_EXTERNAL_AGGRT_ORIGINATED)) {
+		if (!aggr_lsa) {
+			zlog_warn(
+				"%s: Could not refresh/originate %pFX",
+						__func__,
+						&aggr->p);
+			/* Remove the assert later */
+			assert(aggr_lsa);
+			return;
+		}
+
+		external = (struct ospf6_as_external_lsa *)OSPF6_LSA_HEADER_END
+					(aggr_lsa->header);
+		metric = (unsigned long)OSPF6_ASBR_METRIC(external);
+		tag = ospf6_as_external_lsa_get_tag(aggr_lsa);
+		mtype = CHECK_FLAG(external->bits_metric,
+				   OSPF6_ASBR_BIT_E) ? 2 : 1;
+
+		/* Prepare the external_info for aggregator */
+		ospf6_fill_aggr_route_details(ospf6, aggr);
+		rt_aggr = aggr->route;
+		/* If tag/metric/metric-type modified , then re-originate the
+		 * route with modified tag/metric/metric-type details.
+		 */
+		if ((tag != aggr->tag)
+		    || (metric != (unsigned int)rt_aggr->path.cost)
+		    || (mtype != aggr->mtype)) {
+
+			if (IS_OSPF6_DEBUG_AGGR)
+				zlog_debug(
+					"%s: Routetag(old:%d new:%d)/Metric(o:%u,n:%u)/mtype(o:%d n:%d) modified,So refresh the summary route.(%pFX)",
+					__func__, tag, aggr->tag,
+					metric,
+					aggr->metric,
+					mtype, aggr->mtype,
+					&aggr->p);
+
+			aggr_lsa = ospf6_originate_type5_type7_lsas(aggr->route,
+								    ospf6);
+			if (aggr_lsa)
+				SET_FLAG(aggr->aggrflags,
+					OSPF6_EXTERNAL_AGGRT_ORIGINATED);
+		}
+
+		return;
+	}
+
+	/* If the external route prefix same as aggregate route
+	 * and if external route is already originated as TYPE-5
+	 * then it need to be refreshed and originate bit should
+	 * be set.
+	 */
+	if (lsa && prefix_same((struct prefix *)&aggr->p,
+				(struct prefix *)&rt->prefix)) {
+		if (IS_OSPF6_DEBUG_AGGR)
+			zlog_debug("%s: External route prefix is same as aggr so refreshing LSA(%pFX)",
+				__PRETTY_FUNCTION__,
+				&aggr->p);
+
+		THREAD_OFF(lsa->refresh);
```

Comment:
This `THREAD_OFF` is not necessary, `thread_add_event` will figure out there is an event already schedule and won't try to do it again.

Reply:
> This `THREAD_OFF` is not necessary, `thread_add_event` will figure out there is an event already schedule and won't try to do it again.

This is required here because this thread is associated with timer as well, to refresh the LSA periodically. So we need to switch off the timer when we want to refresh the LSA instantly by calling this event. Otherwise the function _thread_add_event won't add the thread into its list considering its already present in the list.